stellargraph/layer/deep_graph_infomax.py

```diff
@@ -0,0 +1,162 @@
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from . import GCN, GAT, APPNP, PPNP
+from ..core.experimental import experimental
+
+from tensorflow.keras.layers import Input, Lambda, Layer, GlobalAveragePooling1D
+import tensorflow as tf
+from tensorflow.keras import backend as K
+import warnings
+
+__all__ = ["DeepGraphInfomax", "DGIDiscriminator"]
+
+
+class DGIDiscriminator(Layer):
+    """
+    This Layer computes the Discriminator function for Deep Graph Infomax (https://arxiv.org/pdf/1809.10341.pdf).
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def build(self, input_shapes):
+
+        first_size = input_shapes[0][-1]
+        second_size = input_shapes[1][-1]
+
+        self.kernel = self.add_weight(
+            shape=(first_size, second_size),
+            initializer="glorot_uniform",
+            name="kernel",
+            regularizer=None,
+            constraint=None,
+        )
+        self.built = True
+
+    def call(self, inputs):
+        """
+        Applies the layer to the inputs.
+
+        Args:
+            inputs: a list or tuple of tensors with shapes [(batch, N, F1), (batch, F2)] containing the node features and a
+                summary feature vector.
+        Returns:
+            a Tensor with shape (1, N)
+        """
+
+        features, summary = inputs
+
+        score = tf.linalg.matvec(features, tf.linalg.matvec(self.kernel, summary),)
+
+        return score
+
+
+@experimental(reason="lack of unit tests", issues=[1003])
+class DeepGraphInfomax:
+    """
+    A class to wrap stellargraph models for Deep Graph Infomax unsupervised training
+    (https://arxiv.org/pdf/1809.10341.pdf).
+
+    Args:
+        base_model: the base stellargraph model class
+    """
+
+    def __init__(self, base_model):
+
+        if not isinstance(base_model, (GCN, GAT, APPNP, PPNP)):
+            raise TypeError(
+                f"base_model: expected GCN, GAT, APPNP or PPNP found {type(base_model).__name__}"
+            )
+
+        if base_model.multiplicity != 1:
+            warnings.warn(
+                f"multiplicity: expected the base_model to have a multiplicity of 1, found"
+                f" ({self.base_model.multiplicity}). A multiplicity of 1 will be used to construct the base model."
+            )
+
+        self.base_model = base_model
+
+        self._node_feats = None
+        self._unique_id = hex(id(self))
```

Comment:
Just to make it look nice when people investigate or visualise the model, maybe we could include a hint about where this came from, e.g.

```suggestion
        self._unique_id = f"DEEP_GRAPH_INFOMAX_{id(self)}"
```

(Since it's building the string directly, no need to hex-ify it, although it could be e.g.`...{id(self):x}`.)